### ISSUE TITLE
Support for KSP

### ISSUE THREAD

**Issue by yigit** [tschuchortdev/kotlin-compile-testing#48]
It would be great to have support for [KSP](https://github.com/android/kotlin/tree/ksp/libraries/tools/kotlin-symbol-processing-api).

I've prototyped it [here](https://github.com/yigit/kotlin-compile-testing/blob/ksp/src/test/kotlin/com/tschuchort/compiletesting/KspProcessorTest.kt)

Before I try to create a proper PR, i wanted to get your opinion on whether this is something you would be interested in getting as a contribution. I'm afraid this may not be very straightforward as it requires depending on 1.4-M1 ([build.gradle file](https://github.com/yigit/kotlin-compile-testing/blame/ksp/build.gradle)).

The project seems to be working fine with 1.4-M1 (except for the regex fix for dependencies) so maybe, it can be a feasible option to create another module that adds extension function to `KotlinCompilation` ?

Something like:

```
fun KotlinCompilation.withKspProcessor(<ksp processor impl>)
```

I've not tried it and not sure if it would work without changing base project to depend on 1.4-M1 though.

**Comment by tschuchortdev** [comment]
Yes, I would be interested in that. 

> I'm afraid this may not be very straightforward as it requires depending on 1.4-M1 (build.gradle file).

What's not straight forward about that? I can only make it a pre-release though until Kotlin 1.4 is out for good because Kotlin-Compile-Testing can only support one compiler version at a time (unless you have an idea how to support multiple versions). 

> it can be a feasible option to create another module that adds extension function to KotlinCompilation ?

A new module and separate artifact is definitely the right way to go, because I don't want to make people depend on KSP when they don't really need it. Should the module be part of this project or part of KSP?

**Comment by yigit** [comment]
I thought it would be better to have it as part of this project for end users' convenience but  i'm afraid they cannot be built together as part of 1 project unless you want to bump kotlin version to 1.4 (which will prevent users in 1.3 so probably not a good idea).

I think there are 3 options:

a) I can create a separate project that depends on this and adds the KSP support. Given that the 1.4 regex fix is merged, i "think" it should work but not sure.

b) same as A but create it in this repository (but still an independent project to avoid bumping kotlin version)

c) Introduce an environment variable like `SUPPORT-KSP` into the build file and put all KSP related code into separate folders. We can optionally include the KSP code based on that environment variable and ship an artifact in addition to the original artifact.

So something like:

```
implementation 'com.github.tschuchortdev:kotlin-compile-testing:1.2.9-withKsp'
```
or
```
implementation 'com.github.tschuchortdev:kotlin-compile-testing-withKsp:1.2.9'
```

I'm OK with any approach you prefer.

**Comment by tschuchortdev** [comment]
Why not implement everything on a separate branch where the Kotlin version can be at 1.4-M1 and release it as a pre-release while the master branch stays at 1.3? As soon as Kotlin 1.4 is released, the KSP support can be merged into master (but still as a separate module/artifact). Hopefully the 1.4 release won't take much longer and if I understand correctly, KSP is not production ready yet.

**Comment by yigit** [comment]
ugh. i didn't think about that 🤦. 

Ok, I'll cleanup my prototype and send a PR for a branch.

The only annoyance would be rebasing that branch before every release. Though "maybe" I can make it work via include build feature of gradle (and keep it a separate build) in which case it could go into master branch as well (and not change the kotlin version).

I'll give it a try and let you know how it goes.
Thanks.

**Comment by tschuchortdev** [comment]
Yes, it is a maintenance burden to keep the branch up-to-date with any changes until the release of Kotlin 1.4 but this project doesn't receive that many commits so it should be doable. The release itself will take the most amount of time because I have to do it manually and the Sonatype Nexus repository never works correctly. 

> "maybe" I can make it work via include build feature of gradle

I'm looking forward to seeing what you come up with.

**Comment by tschuchortdev** [comment]
Resolved with #49
